src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/KeyboardEventSender.cs

```diff
@@ -135,18 +148,46 @@ public static void SendKeyEvent(int replaySegment, Key key, KeyState upOrDown)
             }
         }
 
+        /**
+         * Coroutine to send key events in the given list.
+         * If the same key is encountered multiple times in the list, or the Shift key is encountered, this will
+         * schedule a new event to occur on the next input system update cycle.
+         */
+        public static IEnumerator SendKeyEvents(int replaySegment, IList<(Key, KeyState)> keyStateList)
+        {
+            Dictionary<Key, KeyState> keyStates = new();
+            foreach (var (key, upOrDown) in keyStateList)
+            {
+                if (key == Key.LeftShift || key == Key.RightShift || keyStates.ContainsKey(key))
+                {
+                    if (keyStates.Count > 0)
+                    {
+                        // send the current key states and start a new event after the input system update
+                        SendKeysInOneEvent(replaySegment, keyStates);
+                        yield return WaitForInputSystemUpdate();
+                        keyStates.Clear();
+                    }
+                }
+                keyStates[key] = upOrDown;
+            }
+            if (keyStates.Count > 0)
+            {
+                SendKeysInOneEvent(replaySegment, keyStates);
```

Comment:
I think we might need a `yield return WaitForInputSystemUpdate()` after this line.

If i were to call this method 2x from my calling code, and neither call had duplicated keys in it, but I did have the same keys in call 2 that I had in call 1, we would send 2 events in a single update pass with conflicting keys.

Reply:
Fair point, I will make that change. I now wonder if I should just convert the entire SimpleRandomMovementBehaviour's Update() method to a coroutine, which waits until the SendKeyEvents coroutine finishes?

Reply:
that would cover our current case, but not all cases / future user code patterns

Reply:
I've added the suggestion, and I've also updated SimpleRandomMovementBehaviour to wait until the key event sending coroutine finishes before doing another update step, since this could cause a similar issue as the one you noted if multiple coroutines get scheduled in parallel.